topside/procedures/procedure.py

```diff
@@ -143,18 +144,28 @@ def __init__(self, procedure_id, steps):
         steps: iterable
             An iterable of ProcedureStep objects ordered from first step
             to last step.
+
+        Members 
+        -------
+
+        components: set
+            An unordered collection of components used from each of the 
+            steps from before. Added to set if the action contains a component
         """
         self.procedure_id = procedure_id
         self.steps = {}
         self.step_list = list(steps)
         self.step_id_to_idx = {}
+        self.components = set()
 
         for i, step in enumerate(steps):
             if step.step_id in self.steps:
                 raise ValueError(
                     f'duplicate step ID {step.step_id} encountered in Procedure initialization')
             self.steps[step.step_id] = step
             self.step_id_to_idx[step.step_id] = i
+            if step.action:
+                self.components.add(step.action[0])
```

Comment:
due to the way that action works, sometimes it might not be a tuple.
probably a getter in the other actions, then an if for tuples case

Reply:
basically sometimes it'll be a tuple, sometimes it'll be a "StateChangeAction" so we need to account for that